Add Azure blob storage support

rolanddb opened the pull request:
This PR adds support for Azure Blob Storage

Some code may be refactored since it is common between S3 / Azure / GCS, as discussed in, #124

cricket007 commented on kafka-connect-azure/config/quickstart-schema-source-for-s3.properties:
```diff
@@ -0,0 +1,25 @@
+name=schema-source
+connector.class=io.confluent.connect.storage.tools.SchemaSourceConnector
+tasks.max=1
+topic=s3_topic
```
This isn't S3 😉

cricket007 commented on kafka-connect-azure/docs/changelog.rst:
```diff
@@ -0,0 +1,74 @@
+.. _s3_connector_changelog:
```
Don't need the S3 Connector changelog in the Azure Connector docs, do you?

cricket007 commented on kafka-connect-azure/docs/conf.py:
```diff
@@ -0,0 +1,267 @@
+# -*- coding: utf-8 -*-
+#
+# Kafka Connect S3 documentation build configuration file, created by
```
"S3" in lots of places in this file

cricket007 commented on kafka-connect-azure/docs/configuration_options.rst:
```diff
@@ -0,0 +1,213 @@
+.. _s3_configuration_options:
```
May want to scan over this and rename to Azure related configurations

cricket007 commented on kafka-connect-azure/docs/index.rst:
```diff
@@ -0,0 +1,13 @@
+.. _connect_s3:
```
Update to `azure`

cricket007 commented on kafka-connect-azure/docs/s3_connector.rst:
```diff
@@ -0,0 +1,316 @@
+S3 Connector
```
Update Documentation to say Azure

cricket007 commented on kafka-connect-azure/licenses.html:
```diff
@@ -0,0 +1,339 @@
+</THEAD>
+<TBODY>
+<TR>
+<TD><A HREF="https://aws.amazon.com/sdkforjava">aws-java-sdk-core-1.11.86</A></TD><TD>jar</TD><TD>1.11.86</TD><TD></TD></TR>
```
Azure shouldn't need AWS SDK licenses

cricket007 commented on kafka-connect-azure/pom.xml:
```diff
@@ -0,0 +1,264 @@
+    <dependencies>
+        <dependency>
+            <groupId>io.confluent</groupId>
+            <artifactId>kafka-connect-s3</artifactId>
```
Is this dependency kept until there is a `common-cloud` dependency?

rolanddb replied:
Yes, this was suggested by @kkonstantine here: https://github.com/confluentinc/kafka-connect-storage-cloud/issues/124#issuecomment-344676197

rhauch commented:
@rolanddb Thanks for the submission! It looks like there are failures in the test cases:

```
[INFO] Running io.confluent.connect.AzBlobSinkConnectorIT
[ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.326 s <<< FAILURE! - in io.confluent.connect.AzBlobSinkConnectorIT
[ERROR] testPutRecords(io.confluent.connect.AzBlobSinkConnectorIT)  Time elapsed: 0.319 s  <<< ERROR!
org.apache.kafka.common.config.ConfigException: Configuration avro.codec is defined twice.
	at io.confluent.connect.AzBlobSinkConnectorIT.testPutRecords(AzBlobSinkConnectorIT.java:49)

[ERROR] testStartStop(io.confluent.connect.AzBlobSinkConnectorIT)  Time elapsed: 0.001 s  <<< ERROR!
org.apache.kafka.common.config.ConfigException: Configuration avro.codec is defined twice.
	at io.confluent.connect.AzBlobSinkConnectorIT.testStartStop(AzBlobSinkConnectorIT.java:44)

[INFO] Running io.confluent.connect.AzBlobSinkTaskIT
{azblob.containername=null, format.class=io.confluent.connect.azblob.format.avro.AvroFormat, flush.size=3, schema.generator.class=io.confluent.connect.storage.hive.schema.DefaultSchemaGenerator, azblob.storageaccount.connectionstring=null, store.url=someurl, storage.class=io.confluent.connect.azblob.storage.AzBlobStorage}[2017-12-04 20:11:27,885] INFO Starting SinkTask (io.confluent.connect.azblob.AzBlobSinkTask:97)
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.141 s <<< FAILURE! - in io.confluent.connect.AzBlobSinkTaskIT
[ERROR] put(io.confluent.connect.AzBlobSinkTaskIT)  Time elapsed: 0.138 s  <<< ERROR!
org.apache.kafka.connect.errors.ConnectException: org.apache.kafka.common.config.ConfigException: Configuration avro.codec is defined twice.
	at io.confluent.connect.AzBlobSinkTaskIT.put(AzBlobSinkTaskIT.java:73)
Caused by: org.apache.kafka.common.config.ConfigException: Configuration avro.codec is defined twice.
	at io.confluent.connect.AzBlobSinkTaskIT.put(AzBlobSinkTaskIT.java:73)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   AzBlobSinkConnectorIT.testPutRecords:49 ? Config Configuration avro.codec is d...
[ERROR]   AzBlobSinkConnectorIT.testStartStop:44 ? Config Configuration avro.codec is de...
[ERROR]   AzBlobSinkTaskIT.put:73 ? Connect org.apache.kafka.common.config.ConfigExcepti...
[INFO] 
[ERROR] Tests run: 3, Failures: 0, Errors: 3, Skipped: 0
```

cricket007 commented on kafka-connect-azure/src/main/java/io/confluent/connect/azblob/AzBlobSinkConnectorConfig.java:
```diff
@@ -0,0 +1,328 @@
+package io.confluent.connect.azblob;
+
+import io.confluent.connect.azblob.storage.AzBlobStorage;
+import io.confluent.connect.s3.format.avro.AvroFormat;
```
I see you made a `io.confluent.connect.azblob.format.avro.AvroFormat` class

cricket007 commented on kafka-connect-azure/src/main/java/io/confluent/connect/azblob/format/bytearray/ByteArrayFormat.java:
```diff
@@ -0,0 +1,57 @@
+
+  @Override
+  public SchemaFileReader<AzBlobSinkConnectorConfig, String> getSchemaFileReader() {
+    throw new UnsupportedOperationException("Reading schemas from S3 is not currently supported");
```
"S3"

cricket007 commented on kafka-connect-azure/src/main/java/io/confluent/connect/azblob/format/parquet/ParquetRecordWriterProvider.java:
```diff
@@ -0,0 +1,111 @@
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.apache.parquet.avro.AvroParquetWriter;
+import org.apache.parquet.hadoop.ParquetWriter;
```
I didn't think this project supported Parquet. See issue #26.

rolanddb replied:
I have read that issue. After reading it, it was not completely clear to me what the issues with AvroParquetWriter are. 
We need Parquet support so I decided to give it a shot.

Generally, issue #26 talks about two type of issues
1. Implementation issues ("requires an HDFS configuration") -- I actually found it very easy to implement, the AvroParquetWriter has a builder and it just requires an HDFS path which can be easily constructed from the filename.
2. Runtime issues with delivery -- It works for us, though we only tested in standalone mode. I'm not sure which particular issues we need to watch out for. Also, some issues may be specific for S3 and not Azure blob, since these services are different.

What would you propose to do with Parquet?
- remove support
- do more testing
- something else?

cricket007 replied:
I think the commentary in that comment was referring to needing these

```
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
```

And the constructor for `AvroParquetWriter` that only uses a Path (which is an HDFS Path object, as you mention), is deprecated in favor of the one for the `Configuration` object. 

I really don't know the specifics of the delivery guarantees. That needs to be clarified, but it makes sense to me that HDFS dependencies shouldn't be needed, even if the cloud storage providers may offer HDFS compatibility. Parquet is tightly coupled to the HDFS Configuration, though, so there is no way around that to support the filetype, IMO.

cricket007 commented on kafka-connect-azure/src/main/resources/kafka-connect-s3-version.properties:
```diff
@@ -0,0 +1,17 @@
+#
```
Filename for this has "s3"

cricket007 commented on kafka-connect-azure/src/test/java/io/confluent/connect/AzBlobSinkConnectorTest.java:
```diff
@@ -0,0 +1,171 @@
+        assertEquals(AzBlobSinkTask.class, connector.taskClass());
+    }
+
+//    @Test(expected = ConnectException.class)
```
Should all these commented tests be left in here?

cricket007 commented on kafka-connect-azure/src/test/java/io/confluent/connect/StorageSinkTestBase.java:
```diff
@@ -0,0 +1,97 @@
+import java.util.Map;
+import java.util.Set;
+
+public class StorageSinkTestBase {
```
This looks like it was copied from the `core` module. You can instead add that in the `pom.xml` as a dependency to get this class.

rolanddb replied:
Good one. Done.

cricket007 commented on kafka-connect-azure/src/test/java/io/confluent/connect/Secrets.java:
```diff
@@ -0,0 +1,7 @@
+package io.confluent.connect;
+
+public class Secrets {
```
I'm not seeing immediately seeing where this class is used.

rolanddb replied:
It was used previously in an integration test but I switched to environment variables so I don't accidentally commit secrets. Will remove this file.

rolanddb commented:
I think I've addressed all comments from the review, is there anything I can do to further the progress of this PR?

If the build is still failing on the IT (I can't tell because I can't see the output of Jenkins), this is because the IT requires a real AZ blob storage account. I already excluded running *IT.java from the maven build, so if Jenkins is still attempting to run it, something is wrong.

jpoon commented:
Aside from the Jenkins failures, is there anything that is preventing merging? IOW, if we fix the Jenkins failures, are the maintainers open to merging this PR?

sunray2003 commented:
@jpoon Were you able to fix Jenkins failures, waiting for this feature.

rolanddb commented:
We've been running this in production since I first opened the PR (9 months ago) and it still works fine.

cricket007 commented:
I don't suppose you've gotten the changes made by #187 though?

rolanddb commented:
@cricket007 No, I haven't seen that PR. Thanks for pointing it out!

helayoty commented:
What is the status of this PR? how can we use it?

gilvbp commented:
@rolanddb how can I use it? It's possible to merge with the master? Thanks

SnowDinesh commented:
@rolanddb Any plans to get this merged based on @cricket007 comments? Thanks!

tony810430 commented on kafka-connect-azure/src/main/java/io/confluent/connect/azblob/format/parquet/ParquetRecordWriterProvider.java:
```diff
@@ -0,0 +1,111 @@
+            log.info("Opening record writer for: {}", filename);
+            avroSchema = avroData.fromConnectSchema(schema);
+            writer = AvroParquetWriter
+                .<GenericRecord>builder(hdfsppath)
```
Hi @rolanddb,
The `hdfspath` came from `filename` which is `RecordWriterProvider#getRecordWriter`'s parameter.
It seems that it won't contain any information about azure filesystem keyword and container name.
How do you pass these two things to `AvroParquetWriter`?

tony810430 replied:
Hi @rolanddb,
The `hdfspath` came from `filename` which is `RecordWriterProvider#getRecordWriter`'s parameter.
It seems that it won't contain any information about azure filesystem keyword and container name.
How do you pass these two things to `AvroParquetWriter`?

NathanNam commented:
We have release Azure Blob Storage Sink Connector as a separate connector. Could you try this connector? https://www.confluent.io/hub/confluentinc/kafka-connect-azure-blob-storage

pdavidson100 commented:
We are also very interested in using this connector and would love to see it merged. I'm sure https://www.confluent.io/hub/confluentinc/kafka-connect-azure-blob-storage is excellent, but I assume it will eventually become available under the Enterprise License @NathanNam ?  Unfortunately not everyone is free to use projects under the Confluent Enterprise License. 

I do understand why Confluent might be reluctant to merge this PR, and other similar work, when the company offers a supported closed-source equivalent. If merging to this public repo is not an option, maybe as an alternative could we refactor the `kafka-connect-s3` code to pull out common elements into a `kafka-connect-storage-common` module, and allow reuse by other projects as a dependency? That way @rolanddb could make this connector available in a separate github repo and would still benefit from improvements to `kafka-connect-s3`. 

We are also interested in contributing a GCS connector to this project, but I suspect that will conflict with https://www.confluent.io/hub/confluentinc/kafka-connect-gcs for similar reasons.  If we can refactor this code we would also benefit and be free to release the GCS connector as a separate dependent project.

aakashnshah commented:
With reference to @NathanNam's comment, I will close this PR as now there is a separate connector available that implements Azure Blob Storage support.
